### DISCUSSION TITLE
How to use RTQ with SSG?

### DISCUSSION THREAD

**Discussion by michalvadak** [reduxjs/redux-toolkit#2806] [Q&A]
I can't really get it to work. Should I use `refetchOnMountOrArgChange` as I found in redux-toolkit documentation? Or `revalidation` from Next.js?

**Comment by leosuncin** [comment]
Did you check [SSR with Next.js](https://redux-toolkit.js.org/rtk-query/usage/server-side-rendering) in the docs?

Any of these should work 

```typescript
export const getServerSideProps = wrapper.getServerSideProps(
  (store) => async (context) => {
    store.dispatch(apiName.endpoints.queryNameToRunInSSR.initiate());

    await Promise.all(apiName.util.getRunningOperationPromises());

    return {
      props: {},
    };
  },
);
```

```typescript
export const getStaticProps = wrapper.getStaticProps(
  (store) => async (context) => {
    store.dispatch(apiName.endpoints.queryNameToRunInSSR.initiate());

    await Promise.all(apiName.util.getRunningOperationPromises());

    return {
      props: {},
    };
});
```

Example: https://github.com/leosuncin/goal-app

**Reply by michalvadak** [reply]
I did check it. It also says to setup `refetchOnMountOrArgChange` in tips. But I am getting stale data from the getStaticProps implementation.

**Reply by phryneas** [reply]
@michalvadak you need to create a new store for every single request if you use `next-redux-wrapper`. It should be literally impossible to get stale data from an empty store.

**Reply by michalvadak** [reply]
@phryneas Create a new store? What exactly do you mean? I didn't find it anywhere when looking through the documentation.

**Reply by phryneas** [reply]
that `makeStore` function you pass into the redux-next-wrapper on setup has to be a function that gives you a new store every time, not just return the same store all the time.

**Reply by michalvadak** [reply]
I'm not sure if I understand this. But I implemented it the way I find in the docs.

```

export const makeStore = () =>
  configureStore({
    reducer: {
      app: appReducer,
      [appApi.reducerPath]: appApi.reducer,

      [newsApi.reducerPath]: newsApi.reducer,
    },
    middleware: getDefaultMiddleware => getDefaultMiddleware().concat(appApi.middleware, newsApi.middleware),
  })

// Infer the `RootState` and `AppDispatch` types from the store itself
export type AppStore = ReturnType<typeof makeStore>
export type RootState = ReturnType<AppStore['getState']>
export type AppDispatch = AppStore['dispatch']

export const useAppDispatch = () => useDispatch<AppDispatch>()
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector

export const wrapper = createWrapper<AppStore>(makeStore, { debug: process.env.NODE_ENV === 'development' })

```

**Reply by phryneas** [reply]
That will create a new store each time. Then you have to explain what you mean by `stale data` here.

**Reply by michalvadak** [reply]
Maybe I misunderstood the term `stale data` but I mean that I am getting the "old" data that were provided when the app was initially built and not getting any updated data from the API to the page. 

That's why my initial question was _"Should I use refetchOnMountOrArgChange as I found in redux-toolkit documentation? Or revalidation from Next.js?"_

**Reply by phryneas** [reply]
You probably need to distinguish between SSG and SSR:
SSG will run once at build. SSR will run every time the user wants that page.

If you use SSG, your data will be outdated, yes. Then you will need something like `refetchOnMountOrArgChange` with a time that is reasonable. But then I'd also question why you prepare that data in SSG at all - it will show up outdated and then change to up-to-date. That's probably worse than fetching on the client only.

**Reply by michalvadak** [reply]
I need the data at build time because of SEO. But when I added `refetchOnMountOrArgChange` without `revalidate` (from Next.js), then it didn't work even after the time that I set.

**Reply by phryneas** [reply]
Well, `refetchOnMountOrArgChange` can have all kinds of values. What did you add?

**Reply by michalvadak** [reply]
I tried 60s as well as 30s and after that time I've still got "build" data.

**Reply by phryneas** [reply]
What is the code you used to try that?

**Reply by michalvadak** [reply]
`  const { data, isLoading } = useGetNewsListQuery({ page: 1, perPage }, { refetchOnMountOrArgChange: 60 })`

**Reply by phryneas** [reply]
And SSG was more than 60 seconds in the past, and you looked in the Network tab and do not see a new request made on the client? 
What version of next-redux-wrapper are you using? Does it work if you downgrade to v7? v8 has serious problems at the moment.

**Reply by michalvadak** [reply]
Yes exactly. I am currently on v8. Will try on v7.
